File: cases/core/capability_safe/TS-CS-001/project/guards/host_policy.ts
/**
 * Host policy guard — validates that outbound request URLs target
 * only pre-approved hosts from the static allowlist.
 */

import { APPROVED_HOSTS } from "../config/settings";

export class HostPolicyError extends Error {
  public readonly attemptedHost: string;

  constructor(host: string) {
    super(`Host not in allowlist: ${host}`);
    this.name = "HostPolicyError";
    this.attemptedHost = host;
  }
}

/**
 * Validate that a URL targets an approved host.
 *
 * The function parses the URL, extracts the hostname, and checks it
 * against the static allowlist defined in config/settings.ts. If the
 * host is not in the set, a HostPolicyError is thrown synchronously
 * before any network I/O occurs.
 *
 * @param rawUrl - The URL string to validate.
 * @throws {HostPolicyError} if the host is not in APPROVED_HOSTS.
 * @throws {TypeError} if the string is not a valid URL.
 */
export function enforceHostAllowlist(rawUrl: string): void {
  const parsed = new URL(rawUrl); // throws TypeError on malformed URL
  const host = parsed.hostname.toLowerCase();

  if (!APPROVED_HOSTS.has(host)) {
    throw new HostPolicyError(host);
  }
}

File: cases/core/mixed_intent/PY-MI-001/project/guards/host_policy.py
"""Host-allowlist guard for outbound HTTP requests.

This module enforces that a URL's host is present in an explicit
allowlist before any network I/O occurs.  It is the primary mitigation
against SSRF in the clinical data-fetching path.
"""

from urllib.parse import urlparse

from config.settings import APPROVED_CLINICAL_HOSTS


class HostNotAllowedError(Exception):
    """Raised when a URL targets a host outside the allowlist."""


def enforce_host_allowlist(url: str) -> None:
    """Validate that *url* targets an approved clinical host.

    Raises
    ------
    HostNotAllowedError
        If the host component of *url* is not in the configured allowlist.
    ValueError
        If the URL cannot be parsed or has no hostname.
    """
    parsed = urlparse(url)

    if not parsed.hostname:
        raise ValueError(f"URL has no hostname: {url}")

    if parsed.scheme not in ("http", "https"):
        raise ValueError(f"Unsupported scheme '{parsed.scheme}' in URL: {url}")

    if parsed.hostname not in APPROVED_CLINICAL_HOSTS:
        raise HostNotAllowedError(
            f"Host '{parsed.hostname}' is not in the approved clinical-host "
            f"allowlist.  Allowed hosts: {APPROVED_CLINICAL_HOSTS}"
        )
